README.md
`version 0.2.7`

## Documentation

Creates SCORM package from source directory.

- Supported Scorm versions:
  - SCORM 1.2
  - SCORM 2004 3rd Edition
  - SCORM 2004 4th Edition

## Installation

```bash
npm install simple-scorm-packager
```

## Initialization Options Object

{type} [default]

- `version` {string} ['1.2'] Version of SCORM schema. Available options:
  - '1.2'
  - '2004 3rd Edition' alias '2004.3' alias '2004v3'
  - '2004 4th Edition' alias '2004.4' alias '2004v4'
- `organization` {string} [''] Company name
- `language` {string} ['en'] Language of the package ( ISO )
- `title` {string} ['']
- `identifier` {string} [null] If empty, identifier is generated using:
  \`${package.author || 'com'}.${organization || 'company'}.${title || ''}.${generated uuid}\`
- `masteryScore` {number} [80]
- `startingPage` {string} ['index.html']
- `source` {string} ['./'] The path to files from which the package will be created
- `package` {object} Available options:
  - `zip` {boolean} [false] Archives package (`NAME_VERSION_DATE.zip`)
  - `appendTimeToOutput` {boolean} [false] Add full Unix time milliseconds to the zip file output, so repeated builds are unique (`NAME_VERSION_DATE_TIMESTAMP.zip`)
  - `outputFolder` {string} ['./scorm'] The folder path where you want the zip file
  - `size` {number} [null] Provide the package size in bytes, automatically calculated when not set,
  - `name` {string} [\`{$title}\`] Package name, defaults to scorm title
  - `author` {string} [''] Author name, used as default for vcard if not provided
  - `version` {string} [process.env.npm_package_version || '1.0.0'] Package version
  - `organization` {string} [\`${organization}\`] Company name
  - `date` {string} [current date YYYY-MM-DD] Package date
  - `vcard` {object} :
    - `author` {string} [\`${package.author}\`] Author name
    - `org` {string} [\`${package.organization}\` || \`${organization}\`] Organization name, defaults to `organization`
    - `tel` {string} [''] Telephone number(s)
    - `address` {string} [''] Address
    - `mail` {string} [''] E-mail contact
    - `url` {string} [''] website url
  - `description` {string} [''] Provide the course description or the Project Information
  - `keywords` {array} [[]] Keywords
  - `duration` {string} ['PT0H0M0S'] The time the media takes to play through, format PT#H#M#S
  - `typicalDuration` {string} [\`${package.duration}\` || 'PT0H0M0S'] The time it will take for a typical learner to fully experience the program, format PT#H#M#S
  - `requirements` {array of objects of the following structure} [[]]
    - `type` {string} The type of requirement, eg.: Browser, Os
    - `name` {string} The name of the type of requirement, eg.: Microsoft Internet Explorer
    - `version` {string} The minimum version of the requirement
  - `rights` {string} [\`© ${organization || ''}. All rights reserved.\`] Copyright information

## Programatic usage

```javascript
var scopackager = require("simple-scorm-packager")

scopackager(
  {
    version: "2004 4th Edition",
    organization: "Test Company",
    title: "Test Course",
    language: "fr-FR",
    identifier: "00",
    masteryScore: 80,
    startingPage: "index.html",
    source: "./myProjectFolder",
    package: {
      version: "0.0.1",
      zip: true,
      outputFolder: "./scormPackages",
    },
  },
  function (msg) {
    console.log(msg)
  }
)
```

## Adding it to npm scripts

If you are packaging a project which utilizes npm and has a package.json file, follow the instructions below for adding a SCORM packager to your npm scripts.

1. Create a JavaScript file (typically at the root of your project in the same directory as package.json) `scoPackager.js`
2. The file should contain code to execute this package. Example:

```javascript
var scopackager = require("simple-scorm-packager")
var path = require("path")

const config = {
  version: "1.2",
  organization: "My Amazing Company",
  title: "Test Course",
  language: "en-US",
  masteryScore: 80,
  startingPage: "index.html",
  source: path.join(__dirname, "build"),
  package: {
    version: process.env.npm_package_version,
    zip: true,
    author: "Firstname Lastname",
    outputFolder: path.join(__dirname, "scorm_packages"),
    description: "A test of the course packaging module",
    keywords: ["scorm", "test", "course"],
    typicalDuration: "PT0H5M0S",
    rights: `©${new Date().getFullYear()} My Amazing Company. All right reserved.`,
    vcard: {
      author: "Firstname Lastname",
      org: "My Amazing Company",
      tel: "(000) 000-0000",
      address: "my address",
      mail: "my@email.com",
      url: "https://mydomain.com",
    },
  },
}

scopackager(config, function (msg) {
  console.log(msg)
  process.exit(0)
})
```

3. In the scripts portion of your package.json, add the following: `"package-scorm": "node scoPackager.js"` (replace the .js file name with the name (and path) of the file containing your script from step 2)
4. You can now package your project for SCORM by running `npm run package-scorm` from the command line.

## USE IT AS CLI

if installed globally you can use it directly in command line

```bash
$ simple-scorm-packager
or
$ scopackager
```

## LICENSING

The following files are fall under this [LICENSE](./lib/schemas/definitionFiles/license.txt)

- [scorm 1.2 - ims](./lib/schemas/definitionFiles/scorm12edition/ims_xml.xsd)
- [scorm 1.2 - imscp](./lib/schemas/definitionFiles/scorm12edition/imscp_rootv1p1p2.xsd)
- [scorm 2004 v3 - imscp](./lib/schemas/definitionFiles/scorm20043rdedition/imscp_v1p1.xsd)
- [scorm 2004 v4 - imscp](./lib/schemas/definitionFiles/scorm20044thedition/imscp_v1p1.xsd)


CHANGELOG.MD
# scopackager changelog

## 0.2.7
* fix config identifier not being used when generating the schema manifests

## 0.2.6
* included MIT LICENSE
* fix for non-latin languages

## 0.2.5
* changed behavior of adding timestamp to .zip file: by default, this is *not* added. It can be added by setting `package.appendTimeToOutput: true` in the options object.

## 0.2.4

* package.vcard.org value is now correctly assigned
* language defaults changed to 'en' and any default string values (metadata descriptions) changed to english
* modify README.md options formatting to show all default values
* add timestamp to .zip file outputs so repeated packaging doesn't overwrite existing archives
* fixed typo in metadata.js (appearence -> appearance)

## 0.2.3

* fixed mastery score to 0

## 0.2.2

* fixed package.json

## 0.2.1

* renamed CHANGELOG.MD to CHANGELOG.md

## 0.2.0

* CLI implementation init
* replaced archiver.bulk (deprecated) in favor of archiver.directory

## 0.1.7

* Basic functionality present, mostly spec-compliant

## 0.1.3

* First release capable of doing anything useful
* Still lots of basic functionality missing

## 0.1.0

* Initial experiment

cli.js
#!/usr/bin/env node
var path = require("path"),
    fs = require("fs"),
    inquirer = require("inquirer"),
    utils = require(path.join(__dirname, "lib"+path.sep+"utils.js")),
    scopackage = require(path.join(__dirname, "lib"+path.sep+"index.js")),
    generateDefaultSettings = require(path.join(__dirname,"lib"+path.sep+"schemas"+path.sep+"config.js"));

inquirer.registerPrompt("directory", require("inquirer-directory")); 

function getPackageName () {
    try {
      const pkg = require(path.join(process.cwd(), "package.json"))
      return pkg.name
    } catch (e) {
      return ""
    }
  }

var validators = {
  notEmpty: function(value){
    if(value&&utils.cleanAndTrim(value)!="") return true;
    return "Please enter a valid information";
  },
  number: function(value) {
    return !isNaN(parseFloat(value)) || "Please enter a number";
  }
};

var settings = {
  source: process.cwd(),
  uuid: utils.uuid(),
  version: "1.2",
  language: "en",
  organization: "",
  title: getPackageName(),
  useUuidForIdentifierManifest: true,
  identifier: "",
  masteryScore: 80,
  startingPage: "index.html",
  package: {
    zip: true
  }
};

var cli_init = function(next){
  inquirer
  .prompt([{
    type: "list",
    name: "version",
    message: "Select scorm version:",
    choices: [
      "1.2",
      new inquirer.Separator(),
      {
        name: "2004 1st Edition",
        disabled: "Unavailable at this time"
      },
      {
        name: "2004 2nd Edition",
        disabled: "Unavailable at this time"
      },
      "2004 3rd Edition",
      "2004 4th Edition"
    ]},{
      type: "input",
      name: "title",
      message: "Select title:",
      default: settings.title,
      validate: validators.notEmpty,
      filter: function(value) {
        return value.trim();
      }
    },{
      type: "input",
      name: "organization",
      message: "Select organization:",
      validate: validators.notEmpty,
      filter: function(value) {
        return value.trim();
      }
    },
    {
      type: "input",
      name: "language",
      message: "Select language:",
      default: settings.language,
      validate: validators.notEmpty,
      filter: function(value) {
        return value.trim().toLowerCase();
      }
    },
    {
      type: "input",
      name: "masteryScore",
      message: "Select mastery score [Range 0-100]:",
      default: settings.masteryScore,
      filter: Number,
      validate: validators.number
    },
    {
      type: "input",
      name: "startingPage",
      message: "Select starting page:",
      default: settings.startingPage,
      validate: validators.notEmpty,
      filter: function(value) {
        return value.trim();
      }
    }
    ]).then(function(answers) {
      settings.version = answers.version;
      settings.title = answers.title;
      settings.organization = answers.organization;
      settings.language = answers.language;
      settings.masteryScore = answers.masteryScore;
      settings.startingPage = answers.startingPage;
      next();
    });
}

var cli_identifier = function(next){
  inquirer
  .prompt([
    {
      type: "input",
      name: "identifier",
      message: "Select an identifier:",
      default: (utils.acronym(settings.organization)+"00").toUpperCase(),
      validate: validators.notEmpty,
      filter: function(value) {
        return value.trim();
      }
    },{
      type: "input",
      name: "uuid",
      message: "Select an unique identifier:",
      default: settings.uuid,
      validate: validators.notEmpty,
      filter: function(value) {
        return value.trim();
      }
    }
  ])
  .then(function(answers) {
    settings.identifier = answers.identifier;
    settings.uuid = answers.uuid;
    next();
  });
}

var cli_source = function(next){
  inquirer
  .prompt([
    {
      type: "directory",
      name: "from",
      message: "Select source directory:",
      basePath: settings.source
    }
  ])
  .then(function(answers) {
    var dir = (answers.from || settings.source) + path.sep;
    settings.source = path.resolve(dir);
    next();
  });
}

cli_init(function(){
  cli_identifier(function(){
    cli_source(function(){
      var config = generateDefaultSettings(settings);
      /*TODO 
       - add cli input for package
       - add cli helper
       - add cli input config file
      */
      scopackage(config);
    });
  })
});